[inductor] Fix fresh_inductor_cache()

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122661

Summary: Modify fresh_inductor_cache() to clear cached state before mocking the toplevel cache_dir directory. Any lru_caches (or otherwise) can use the @clear_on_fresh_inductor_cache decorator to register the cache for clearing. Also change the base inductor TestCase class to use fresh_inductor_cache(). Previously that TestCase was only mocking the subdirectory within the toplevel cache dir designated for the FX graph cache artifacts.

Test Plan:
- New unit test
- All existing inductor tests will exercise fresh_inductor_cache()

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

masnesral committed:
[inductor] Fix fresh_inductor_cache()

Summary: Modify fresh_inductor_cache() to clear cached state before mocking the toplevel cache_dir directory. Any lru_caches (or otherwise) can use the @clear_on_fresh_inductor_cache decorator to register the cache for clearing. Also change the base inductor TestCase class to use fresh_inductor_cache(). Previously that TestCase was only mocking the subdirectory within the toplevel cache dir designated for the FX graph cache artifacts.

Test Plan:
- New unit test
- All existing inductor tests will exercise fresh_inductor_cache()

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Fix fresh_inductor_cache()"


Summary: Modify fresh_inductor_cache() to clear cached state before mocking the toplevel cache_dir directory. Any lru_caches (or otherwise) can use the clear_on_fresh_inductor_cache decorator to register the cache for clearing. Also change the base inductor TestCase class to use fresh_inductor_cache(). Previously that TestCase was only mocking the subdirectory within the toplevel cache dir designated for the FX graph cache artifacts.

Test Plan:
- New unit test
- All existing inductor tests will exercise fresh_inductor_cache()

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -167,6 +172,7 @@ def get_system() -> Dict[str, Any]:
         return system
 
     @staticmethod
+    @clear_on_fresh_inductor_cache
```
I did not try to modify the behavior of `get_global_cache_path` below since it seems like a different thing altogether, but lemme know if this is something that should also reset with `fresh_inductor_cache`

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -1593,9 +1599,10 @@ def split_aot_inductor_output_path(path: str) -> Tuple[str, str]:
 class CudaKernelParamCache:
     cache: Dict[str, Dict[str, str]] = dict()
-    clear = staticmethod(cache.clear)
+    cache_clear = staticmethod(cache.clear)
```
In order to use the new decorator, the clear method must be called `cache_clear`. I assume it's ok to change the names since I found no references in the existing code base; nothing here is a public API right?

oulgen replied:
This is under _inductor, everything here is private api

oulgen commented on torch/_inductor/utils.py:
```diff
@@ -727,6 +727,22 @@ def instance_descriptor(
+    Use this decorator to register any caches that should be cache_clear'd
+    with fresh_inductor_cache().
+    """
+    if not hasattr(obj, "cache_clear") or not callable(obj.cache_clear):
+        raise AttributeError(f"{obj} does not have a cache_clear method")
```
do we need this? dont we only add this to "correct" functions?

masnesral replied:
Yes, it's not strictly needed. I'm just trying to give a more readable error message if someone slaps this decorator on a function incorrectly. Ideally, I could get mypy to complain if used incorrectly, but AFAIK there's no way to do it robustly because Python.

oulgen reviewed (approved):
Looks great, thanks!

masnesral commented:
@JackCaoG I have a test failure for //test/cpp:test_aten_xla_tensor_6 that I'm 99% sure can't be related to this change, but would you mind verifying?

JackCaoG commented:
You can ignore the XLA test failure, we are trying to update the XLA pin in https://github.com/pytorch/pytorch/pull/122628 to fix the test failure.

masnesral commented:
@pytorchbot merge -i

masnesral commented:
@pytorchbot merge -i

oulgen commented:
@masnesral you seem to have failing test
> inductor/test_triton_heuristics.py::TestTritonHeuristics::test_artificial_zgrid

not the XLA failure

masnesral commented:
> you seem to have failing test

Yeah, that one fails on the parent for me too (and on main), so I believed it was unrelated to my change.

masnesral committed:
rebase on "[inductor] Fix fresh_inductor_cache()"


Summary: Modify fresh_inductor_cache() to clear cached state before mocking the toplevel cache_dir directory. Any lru_caches (or otherwise) can use the clear_on_fresh_inductor_cache decorator to register the cache for clearing. Also change the base inductor TestCase class to use fresh_inductor_cache(). Previously that TestCase was only mocking the subdirectory within the toplevel cache dir designated for the FX graph cache artifacts.

Test Plan:
- New unit test
- All existing inductor tests will exercise fresh_inductor_cache()

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison reviewed (commented):
letting @oulgen take review on this one but looks good!

masnesral committed:
rebase on "[inductor] Fix fresh_inductor_cache()"


Summary: Modify fresh_inductor_cache() to clear cached state before mocking the toplevel cache_dir directory. Any lru_caches (or otherwise) can use the clear_on_fresh_inductor_cache decorator to register the cache for clearing. Also change the base inductor TestCase class to use fresh_inductor_cache(). Previously that TestCase was only mocking the subdirectory within the toplevel cache dir designated for the FX graph cache artifacts.

Test Plan:
- New unit test
- All existing inductor tests will exercise fresh_inductor_cache()

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Fix fresh_inductor_cache()"


Summary: Modify fresh_inductor_cache() to clear cached state before mocking the toplevel cache_dir directory. Any lru_caches (or otherwise) can use the clear_on_fresh_inductor_cache decorator to register the cache for clearing. Also change the base inductor TestCase class to use fresh_inductor_cache(). Previously that TestCase was only mocking the subdirectory within the toplevel cache dir designated for the FX graph cache artifacts.

Test Plan:
- New unit test
- All existing inductor tests will exercise fresh_inductor_cache()

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "new test is failing internally" -c ghfirst

int3 commented:
Since this has been reverted -- could you add a flag / env var to disable cache deletion before landing it again? Having the cache deleted makes debugging failing tests a lot more inconvenient

masnesral committed:
Update on "[inductor] Fix fresh_inductor_cache()"


Summary: Modify fresh_inductor_cache() to clear cached state before mocking the toplevel cache_dir directory. Any lru_caches (or otherwise) can use the clear_on_fresh_inductor_cache decorator to register the cache for clearing. Also change the base inductor TestCase class to use fresh_inductor_cache(). Previously that TestCase was only mocking the subdirectory within the toplevel cache dir designated for the FX graph cache artifacts.

Test Plan:
- New unit test
- All existing inductor tests will exercise fresh_inductor_cache()

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented on torch/_inductor/test_case.py:
```diff
@@ -20,34 +20,13 @@ class TestCase(DynamoTestCase):
-        mock_get_dir.return_value = self._inductor_cache_tmp_dir.name
+        self._inductor_test_stack = contextlib.ExitStack()
+        self._inductor_test_stack.enter_context(config.patch({"fx_graph_cache": True}))
+        if os.environ.get("INDUCTOR_TEST_DISABLE_FRESH_CACHE") != "1":
```
@int3 how's this env var?

int3 replied:
That seems good, thank you!

masnesral committed:
rebase on "[inductor] Fix fresh_inductor_cache()"


Summary: Modify fresh_inductor_cache() to clear cached state before mocking the toplevel cache_dir directory. Any lru_caches (or otherwise) can use the clear_on_fresh_inductor_cache decorator to register the cache for clearing. Also change the base inductor TestCase class to use fresh_inductor_cache(). Previously that TestCase was only mocking the subdirectory within the toplevel cache dir designated for the FX graph cache artifacts.

Test Plan:
- New unit test
- All existing inductor tests will exercise fresh_inductor_cache()

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented:
I see accuracy failures on many other PRs for `torch_multimodal_clip` on this same job, so pretty sure this change is unrelated.

masnesral commented:
@pytorchbot merge -i

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
Hi @masnesral! Sorry, I need to revert your PR because it breaks the internal build. Here you can find more information about the issue - [D56012761](https://www.internalfb.com/diff/D56012761).

oulgen commented:
It looks like a pyre typechecking failed

> [deeplearning/aot_inductor/test/test_lower.py:114:12] Unused ignore [0]: The `pyre-ignore[58]` or `pyre-fixme[58]` comment is not suppressing type errors, please remove it.

which feels unrelated?

masnesral commented:
> It looks like a pyre typechecking failed

yeah, that's pretty confusing. I can't see how it has anything to do with this PR. But pyre _does not_ fail on the parent, so ¯\_(ツ)_/¯

oulgen commented:
@masnesral I'd put the diff on fbcode and try landing it directly there, and let the automation land it here on github

masnesral commented:
@oulgen Yeah, the pyre-fixme looks legit and seems like it should still complain even after the changes in this PR. Something here is fooling pyre into relaxing the typecheck. Looks like I'll need to fix some merge conflicts anyway. I put up D56142358 internally that I _think_ should get me past the problem.

masnesral committed:
rebase+merge on "[inductor] Fix fresh_inductor_cache()"


Summary: Modify fresh_inductor_cache() to clear cached state before mocking the toplevel cache_dir directory. Any lru_caches (or otherwise) can use the clear_on_fresh_inductor_cache decorator to register the cache for clearing. Also change the base inductor TestCase class to use fresh_inductor_cache(). Previously that TestCase was only mocking the subdirectory within the toplevel cache dir designated for the FX graph cache artifacts.

Test Plan:
- New unit test
- All existing inductor tests will exercise fresh_inductor_cache()

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
